src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -1326,6 +1329,126 @@ public InMemoryOutput downloadOutputs(RemoteAction action, RemoteActionResult re
+      spawnResultFuture.cancel(true);
+    }
+  }
+
+  /**
+   * @return Whether the spawn result should be uploaded to the cache.
+   */
+  public boolean shouldStore(SpawnResult result, @Nullable LocalExecution execution) {
+    if (SpawnResult.Status.SUCCESS.equals(result.status()) && result.exitCode() == 0) {
+      if (execution != null) {
+        execution.spawnResultFuture.set(result);
+      }
+      return true;
+    } else {
+      if (execution != null) {
+        execution.spawnResultFuture.cancel(true);
+      }
+      return false;
+    }
+  }
+
+  /**
+   * Reuses the outputs of a concurrent local execution of the same RemoteAction in a different
+   * spawn.
+   *
+   * <p>Since each output file is generated by a unique action and action's generally take care to
+   * run a unique spawn for each output file, this method is only useful with path mapping enabled,
+   * which allows different spawns in a single build to have the same RemoteAction.ActionKey.
+   */
+  @Nullable
+  public SpawnResult waitForAndReuseOutputs(RemoteAction action, LocalExecution previousExecution)
+      throws InterruptedException, IOException {
+    checkState(!shutdown.get(), "shutdown");
+
+    SpawnResult previousSpawnResult;
+    try {
+      previousSpawnResult = previousExecution.spawnResultFuture.get();
+    } catch (CancellationException | ExecutionException e) {
+      Throwables.propagateIfPossible(e.getCause(), InterruptedException.class);
+      // The spawn this action was deduplicated against failed due to an exception or
+      // non-zero exit code. Since it isn't possible to transparently replay its failure for the
+      // current spawn, we rerun the action instead.
+      return null;
+    }
+
+    Preconditions.checkArgument(
+        action.getActionKey().equals(previousExecution.action.getActionKey()));
+
+    Map<Path, ActionInput> previousOutputs =
+        previousExecution.action.getSpawn().getOutputFiles().stream()
+            .collect(toImmutableMap(output -> execRoot.getRelative(output.getExecPath()), o -> o));
+    Map<Path, Path> realToTmpPath = new HashMap<>();
+    for (String output : action.getCommand().getOutputPathsList()) {
+      Path sourcePath =
+          previousExecution
+              .action
+              .getRemotePathResolver()
+              .outputPathToLocalPath(encodeBytestringUtf8(output));
+      ActionInput outputArtifact = previousOutputs.get(sourcePath);
+      Path tmpPath = tempPathGenerator.generateTempPath();
+      tmpPath.getParentDirectory().createDirectoryAndParents();
+      try {
+        if (outputArtifact.isDirectory()) {
+          tmpPath.createDirectory();
+          FileSystemUtils.copyTreesBelow(sourcePath, tmpPath, Symlinks.NOFOLLOW);
+        } else if (outputArtifact.isSymlink()) {
+          FileSystemUtils.ensureSymbolicLink(tmpPath, sourcePath.readSymbolicLink());
+        } else {
+          FileSystemUtils.copyFile(sourcePath, tmpPath);
+        }
```

Comment:
FYI - I suspect that this has a chance of failing if we ever apply path mapping to test actions with multiple test attempts (which move spawn outputs to a different location after spawn execution). This is already a problem for async uploading, so whatever we end up doing to fix it should also apply here. No need to do anything about it for now.

Reply:
That's a good catch. At least in its current form, path mapping for test actions wouldn't be very useful as top-level targets that differ in configuration typically do so intentionally, with at least some part of their transitive closure being different (as in, different file content). In this case, path mapping wouldn't help.

Reply:
Yeah, agreed.